cli/src/git_util.rs

```diff
@@ -326,6 +332,150 @@ pub fn print_git_import_stats(
     Ok(())
 }
 
+pub struct Progress {
+    next_print: Instant,
+    rate: RateEstimate,
+    buffer: String,
+    guard: Option<CleanupGuard>,
+}
+
+impl Progress {
+    pub fn new(now: Instant) -> Self {
+        Self {
+            next_print: now + crate::progress::INITIAL_DELAY,
+            rate: RateEstimate::new(),
+            buffer: String::new(),
+            guard: None,
+        }
+    }
+
+    pub fn update<W: std::io::Write>(
+        &mut self,
+        now: Instant,
+        progress: &git::Progress,
```

Comment:
nit: might be better to extract `git::Progress` type to somewhere else. The implementation looks general for file downloading/uploading workload.

Reply:
I agree with this, but for now it's only used in `git_util` anyway, so I think inlining it here is fine as it is right now. I can extract it back out into `progress` later, since I might need it anyway.

Reply:
To minimize the change, I would rather not move codes back and forth. I don't have strong feeling, though.

Reply:
Do we think it will be reused by non-git commands? I kind of feel like it won't.

Reply:
If we add other backend, it might be. For the moment, no, I don't think it will be reused.

Reply:
I mean that even if we add another backend, the current struct (`bytes_downloaded: Option<u64>, overall: f32`) seems too specific to how Git does things that I'm not sure we will reuse it. Perhaps it can be made into a more generic progress bar, however.

Reply:
Yes, the current data struct might have to be adjusted. My feeling is that the UI impl is mostly backend agnostic.